mlir/lib/Target/IRDLToCpp/IRDLToCpp.cpp

```diff
@@ -0,0 +1,530 @@
+                                                 "  odsState.addOperands({0});",
+                                                 attr);
+                                           }),
+                           "\n"),
+                llvm::join(llvm::map_range(opStrings.opResultNames,
+                                           [](StringRef attr) -> std::string {
+                                             return llvm::formatv(
+                                                 "  odsState.addTypes({0});",
+                                                 attr);
+                                           }),
+                           "\n"));
+            dict["OP_BUILD_DEFS"] = buildDefinition;
+
+            std::string str;
+            llvm::raw_string_ostream stream{str};
+            perOpDefTemplate.render(stream, dict);
+            return str;
+          }),
+      "\n");
+
+  dict["OP_LIST"] = commaSeparatedOpList;
+  dict["OP_CLASSES"] = perOpDefinitions;
+  opDefTemplate.render(output, dict);
+  dialectDefTemplate.render(output, dict);
+
+  output << "#endif // " << definitionMacroFlag << "\n";
+  return success();
+}
+
+static LogicalResult verify(irdl::DialectOp dialect) {
+  auto &dialectBlock = getDialectBlock(dialect);
+  for (auto operation : dialectBlock.getOps<irdl::OperationOp>()) {
+
+    for (auto operands : operation.getOps<irdl::OperandsOp>()) {
+      for (auto operand : operands.getOperands()) {
+        if (!llvm::isa<irdl::AnyOp>(operand.getDefiningOp())) {
+          return operands.emitError(
+              "IRDL C++ translation only supports irdl.any "
+              "constraint for types");
+        }
+      }
+    }
+
+    for (auto results : operation.getOps<irdl::ResultsOp>()) {
+      for (auto operand : results.getOperands()) {
+        llvm::errs() << "HELP";
+        if (!llvm::isa<irdl::AnyOp>(operand.getDefiningOp())) {
+          return results.emitError(
+              "IRDL C++ translation only supports irdl.any "
+              "constraint for types");
+        }
+      }
+    }
+  }
+
+  return success();
+}
+
+LogicalResult irdl::translateIRDLDialectToCpp(irdl::DialectOp dialect,
+                                              raw_ostream &output) {
+  static const auto typeDefTempl = detail::Template(
+#include "Templates/TypeDef.txt"
+  );
+
+  StringRef dialectName = dialect.getSymName();
+
+  // TODO: deal with no more constraints than the verifier allows.
+  if (dialectName.size() < 1)
+    return dialect->emitError("dialect name must be more than one character");
+  if (!llvm::isAlpha(dialectName[0]))
+    return dialect->emitError("dialect name must start with a letter");
+  if (!llvm::all_of(dialectName,
+                    [](char c) { return llvm::isAlnum(c) || c == '_'; }))
+    return dialect->emitError(
+        "dialect name must only contain letters, numbers or underscores");
+
+  if (failed(verify(dialect)))
+    return failure();
+
+  // TODO: allow more complex path.
```

Comment:
We may drop this TODO. It's questionable whether the feature is desired in thee first place.